pkg/store/query.go

```diff
@@ -0,0 +1,202 @@
+	logger    log.Logger
+	stores    func() []*query.StoreInfo
+	extLabels labels.Labels
+}
+
+// NewQueryStore returns a new QueryStore that uses the given clients that implements storeAPI to fan-in all series to the client.
+// This store is meant to be used for query federation, thus does not have support for deduplication. Deduplication
+// should be done on the highest level (just before PromQL)
+// It attaches the provided external labels to all results.
+func NewQueryStore(
+	logger log.Logger,
+	stores func() []*query.StoreInfo,
+	extLabels labels.Labels,
+) *QueryStore {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	q := &QueryStore{
+		logger:    logger,
+		stores:    stores,
+		extLabels: extLabels,
+	}
+	return q
+}
+
+// Info returns store information about the external labels this store have.
+func (q *QueryStore) Info(ctx context.Context, r *storepb.InfoRequest) (*storepb.InfoResponse, error) {
+	res := &storepb.InfoResponse{
+		Labels: make([]storepb.Label, 0, len(q.extLabels)),
+	}
+	for _, l := range q.extLabels {
+		res.Labels = append(res.Labels, storepb.Label{
+			Name:  l.Name,
+			Value: l.Value,
+		})
+	}
+	return res, nil
+}
+
+// seriesSet returns all series for a requested time range and label matcher. Requested series are taken from other
+// stores and proxied to RPC client.
+func (q *QueryStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesServer) error {
+	match, newMatchers, err := extLabelsMatches(q.extLabels, r.Matchers)
+	if err != nil {
+		return status.Error(codes.InvalidArgument, err.Error())
+	}
+	if !match {
+		return nil
+	}
+
+	var (
+		respCh = make(chan *storepb.SeriesResponse)
+		g      errgroup.Group
+	)
+	for _, s := range q.stores() {
+		// We might be able to skip the store if its meta information indicates
+		// it cannot have series matching our query.
+		// NOTE: all matchers are validated in extLabelsMatches method so we explicitly ignore error.
+		if ok, _ := storeMatches(s, r.MinTime, r.MaxTime, newMatchers...); !ok {
+			continue
+		}
+		store := s
+
+		g.Go(func() error {
+			sc, err := store.Client.Series(srv.Context(), &storepb.SeriesRequest{
+				MinTime:  r.MinTime,
+				MaxTime:  r.MaxTime,
+				Matchers: newMatchers,
+			})
+			if err != nil {
+				respCh <- storepb.NewWarnSeriesResponse(errors.Wrap(err, "fetch series"))
+				return nil
+			}
+
+			for {
+				r, err := sc.Recv()
+				if err == io.EOF {
+					break
+				}
+				if err != nil {
+					respCh <- storepb.NewWarnSeriesResponse(errors.Wrap(err, "receive series"))
+					return nil
+				}
+
+				if w := r.GetWarning(); w != "" {
+					respCh <- storepb.NewWarnSeriesResponse(errors.New(w))
+					continue
+				}
+
+				series := r.GetSeries()
+				series.Labels = extendLset(series.Labels, q.extLabels)
+				respCh <- storepb.NewSeriesResponse(series)
+			}
+
+			return nil
+		})
+	}
+
+	go func() {
+		_ = g.Wait()
+		close(respCh)
+	}()
+
+	for resp := range respCh {
+		if err := srv.Send(resp); err != nil {
```

Comment:
Is `Send()` not thread safe or why do we need the extra channel?
Regardless, this does not seem to guarantee that we are returning series in order as the streaming API contract requires. We need a merge procedure that aggregates it into a sorted stream.

Reply:
Implementing a series set on top of the incoming stream for a store and wrapping those into [`storepb.MergedSeriesSet`](https://github.com/improbable-eng/thanos/blob/master/pkg/store/storepb/custom.go#L54) should do the trick.

Reply:
TBH I just thought from the beginning that gRPC stream does not guarantee ordered delivery.. but when I read through web... it is (: I wanted to do simple proxy that we send immidiately when we got something, but I guess we cannot go around some kind of aggregation after "fetch" stage. ):

Reply:
The order guarantee rather refers to our application level contract. We always want to send series in order onto the channel. So if we aggregate, we should do the same.

Reply:
done